docs/spec/distributing.rst

```diff
@@ -10,72 +10,428 @@ Stub files
+  formats are ``# type: ignore``, ``# type: ignore[error-class]``, and
+  ``# pyright: ignore[error-class]``. Type checkers may ignore error
+  suppressions that they don't support but should not error on them.
+
+Imports
+"""""""
+
+Stub files distinguish between imports that are re-exported and those
+that are only used internally. See :ref:`import-conventions`.
+
+Type aliases can be used to re-export an import under a different name::
+
+    from foo import bar as _bar
+    new_bar = _bar  # "bar" gets re-exported with the name "new_bar"
+
+Stubs support customizing star import semantics by defining a module-level
+variable called ``__all__``. In stubs, this must be a string list literal.
+Other types are not supported. Neither is the dynamic creation of this
+variable (for example by concatenation).
+
+When ``__all__`` is defined, exactly those names specified in ``__all__`` are
+imported::
+
+    __all__ = ['public_attr', '_private_looking_public_attr']
+
+    public_attr: int
+    _private_looking_public_attr: int
+    private_attr: int
+
+Type checkers support cyclic imports in stub files.
+
+Built-in Generics
+"""""""""""""""""
+
+:pep:`585` built-in generics are supported and should be used instead
+of the corresponding types from ``typing``::
+
+    from collections import defaultdict
+
+    def foo(t: type[MyClass]) -> list[int]: ...
+    x: defaultdict[int]
+
+Using imports from ``collections.abc`` instead of ``typing`` is
+generally possible and recommended::
 
-* Extension modules
+    from collections.abc import Iterable
 
-* Third-party modules whose authors have not yet added type hints
+    def foo(iter: Iterable[int]) -> None: ...
 
-* Standard library modules for which type hints have not yet been
-  written
+Unions
+""""""
+
+Declaring unions with the shorthand syntax or ``Union`` and ``Optional`` is
+supported by all type checkers::
+
+    def foo(x: int | str) -> int | None: ...  # recommended
+    def foo(x: Union[int, str]) -> Optional[int]: ...  # ok
+
+Module Level Attributes
+"""""""""""""""""""""""
+
+Module level variables and constants can be annotated using either
+type comments or variable annotation syntax::
+
+    x: int  # recommended
+    x: int = 0
+    x = 0  # type: int
+    x = ...  # type: int
+
+The ellipsis literal can stand in for any value::
+
+    x: int = ...  # type is int
+
+A variable annotated as ``Final`` and assigned a literal value has the
+corresponding ``Literal`` type::
+
+    x: Final = 0  # type is Literal[0]
+
+In all other cases, the type of a variable is unspecified when the variable is
+unannotated or when the annotation and the assigned value disagree::
+
+    x = 0  # type is unspecified
+    x = ...  # type is unspecified
+    x: int = ""  # type is unspecified
+
+Classes
+"""""""
+
+Class definition syntax follows general Python syntax, but type checkers
+are only expected to understand the following constructs in class bodies:
```

Comment:
Grammar nit: "only expected to understand" should be "expected to understand only"